Hierarchical terms sorted by name and filterable

notnownikki opened the pull request:
## Description

Fix for https://github.com/WordPress/gutenberg/issues/2607

Limits the height of the list so that sites with a lot of categories don't have lists that cause the page to grow a lot.

Adds a filter field to do simple substring matching on the list.

Changes the default ordering to name ascending.

## How has this been tested?

Load a post, check the categories are there. Try filtering the list of categories to find the one you want.

## Screenshots <!-- if applicable -->

![newcategoryui](https://user-images.githubusercontent.com/3314354/45958198-b72c2980-c00e-11e8-8532-ed00a5f2d688.gif)


## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

notnownikki committed:
Hierarchical terms sorted by name and filterable

notnownikki commented:
Looking for accessibility feedback here especially!

notnownikki committed:
Better type for the filter input

afercia commented on packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:
```diff
@@ -268,10 +282,24 @@ class HierarchicalTermSelector extends Component {
 		const inputId = `editor-post-taxonomies__hierarchical-terms-input-${ instanceId }`;
+		const filterInputId = `editor-post-taxonomies__hierarchical-terms-filter-${ instanceId }`;
 
 		/* eslint-disable jsx-a11y/no-onchange */
```
this can be removed as `jsx-a11y/no-onchange` applies only to `<select>` elements

afercia commented on packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:
```diff
@@ -268,10 +282,24 @@ class HierarchicalTermSelector extends Component {
 		/* eslint-disable jsx-a11y/no-onchange */
 		return [
-			...this.renderTerms( availableTermsTree ),
+			<input
```
the search input needs a visible `<label>` element, properly associated with for / id attributes

afercia commented on packages/editor/src/components/post-taxonomies/style.scss:
```diff
@@ -1,3 +1,8 @@
+.editor-post-taxonomies__hierarchical-terms-list {
+	max-height: 14em;
+	overflow: auto;
```
scrollable divs need to be made keyboard accessible: only Firefox makes them focusable by default so for other browsers it's necessary to make them focusable and add some ARIA. See for example the Inserter. I'd say something like `tabIndex="0" role="group" aria-label={ __( 'Available terms' ) }"` could work

afercia commented on packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js:
```diff
@@ -214,9 +215,22 @@ class HierarchicalTermSelector extends Component {
+					( childTerm ) => -1 !== childTerm.name.toLowerCase().indexOf( filterValue.toLowerCase() )
+				).reduce(
+					( appeared, appears ) => appeared || appears
+				) )
```
Seems all this will work only for children. What about grandchildren and deeper nested terms?

afercia commented:
Great to see exploration on this. I've left some comments, focussed on a11y and functionality. Will leave other considerations to the expertise of the GTeam.

In addition to the comments, some more considerations:
- there's no audible feedback about the results of the filtered search: all the other searches / completers in Gutenberg also use `speak` to send an audible message to assistive technologies with the number of results or "no results"
- when there are no results, in addition to the audible message, should some text be used to communicate "no results" also visually?
- as said in the comments, seems it doesn't work for terms nested deeper than 1 level: I'm using the themes unit test data and I have grandchild categories:

<img width="320" alt="screen shot 2018-09-24 at 19 19 08" src="https://user-images.githubusercontent.com/1682452/45968901-8362fb00-c032-11e8-9a25-811e747c2a9e.png">

typing "grandchild" doesn't return any result though:

<img width="331" alt="screen shot 2018-09-24 at 19 18 52" src="https://user-images.githubusercontent.com/1682452/45968921-8eb62680-c032-11e8-9087-b561bcdf5ffd.png">

- say I have only 4-5 categories: should the search field be displayed? the categories will be all displayed because they don't exceed the max-height and a filter functionality seems a bit pointless in this case

notnownikki commented:
@afercia thank you! I think I can take care of all of that :)

notnownikki committed:
Order results by selected or child selected status

notnownikki committed:
Accessibility fixes

notnownikki commented:
@afercia I _think_ I've addressed the concerns now, although I'm not sure how to test the speech. I get the feeling it might need to be debounced, but I'm not sure.

afercia commented:
@notnownikki thanks! Oh yes it needs to be debounced otherwise a message is sent at any keystroke. I think `withSpokenMessages` already exposes both a `speak` and a `debouncedSpeak` prop. See other usages in the codebase.

notnownikki committed:
Debounce speak

notnownikki commented:
Aha, so it does! Ok, using the debounced version now.

afercia commented:
> not sure how to test the speech

- use a screen reader :)
- observe the change in the content of the `wp-a11y-speak-region` div element
- there used to be a plugin to log in the console the `speak` messages, built by @westonruter not sure if it works with the new version of `speak`

karmatosed reviewed (approved):
Design wise to me this works as a start to fix this problem. I am torn on suggesting more padding top/bottom but we can iterate. Thanks for this improvement.

notnownikki commented:
Awesome! @karmatosed thank you so much for your guidance on this :heart:

karmatosed commented:
Thanks for all your work!

mtias commented:
This is great, thanks for the improvement. What do you think about not showing the search box if there are less than 8 items or so?

cc @jasmussen @karmatosed @melchoyce @kjellr @alexislloyd

jasmussen commented:
I think that seems good. Want to make sure removing the search box doesn't mess with the tabbing.
